### ISSUE TITLE
Tweening and transforms

### ISSUE THREAD

**Issue by focustense** [focustense/StardewUI#59]
This is more of a placeholder/brainstorming space than a real feature plan, but inspired by recent sandboxing, I think there could be a place for this to become a first-class feature.

Without having to build all of CSS, it would have to look very different from transitions/animations in web apps. That said, I think it is still possible to design in a way that's both amenable to data binding, but doesn't require a whole bunch of animation and timing logic to go directly into the model/data.

For example, say we define a `Tween` data type with the usual string and duck typing conversion, along with one or more structural attributes and some type of microsyntax:

```html
<image sprite={@...} opacity="0.1" scale="0" *tween="opacity 500ms -> 1.0, scale 250ms -> 1.0" />
```

This is essentially an instruction to _start_ the opacity and scale at 0.1 and 0.0, respectively; animate opacity up to 1.0 over 500 ms, and scale to 1.0 over 250 ms. Being a structural attribute, `*tween* could itself be data-bound, so starting a new animation means just changing that attribute. The hardcoded version above is a one-time tween that would happen when the image is first displayed.

A possibly cleaner and more extensible way would be:

```html
<image sprite={@...} opacity="0.1" *tween-opacity="250ms, 100ms -> 0.8, 200ms -> 1.0" />
```

Meaning "delay for 250 ms at current value, then animate to 0.8 over 100 ms, then animate to 1.0 over 200 ms". Entire animation "curves" can be represented this way, and keeping it separated by property adds more work for the lexer/parser/binding side but is likely to be more convenient to write and maintain, and avoids overly long strings in the markup.

Tweening would always start at the _current_ property value, not whatever the value's literal/binding attribute says, and always animates to the tweened value, again regardless of the attribute. The original attribute becomes just an initial condition, except that we do need to handle changes, so probably if the non-tweened attribute itself changes, we use that to cancel any tweening and immediately update the UI, on the assumption that if the user wanted it to be a tweened transition, they would update the `*tween` instead.

One thing that's missing to make this a broadly useful feature is transforms, and especially transform propagation. `Transform` in `ISpriteBatch` is a cheat, just tracking the translation value; MonoGame/XNA does support matrix transforms, but in a god-awful way, as a parameter to `SpriteBatch.Begin`, and I really don't think we should be starting a new sprite batch for every view, or every `ISpriteBatch.Transform` because there's a lot more overhead than just updating a matrix.

So, not sure how to make it work, but for tweening to be a truly useful feature, views ought to have a layout-independent transform with the standard properties: translation, rotation and scale, these being separate and distinct from the layout properties and view-specific properties like `Image` rotation.

Maybe it's possible to come up with some ugly hack where we do `SpriteBatch.Begin` whenever a "difficult" transform is encountered, mainly rotation, and for translation/scale transforms we just fiddle with the argument values passed into the methods instead including subsequent `Transform`. I think something like this would involve a "base matrix", which explicit translation offsets and scale adjustments could be applied on top of, and then whenever we need a "derived matrix", we have to bake in all the translations and scales applied since that base.

And that still does not solve the problem of `DelegateDraw` which could ignore all transforms; on the other hand, maybe it actually does provide a better path to solving that problem, because we can compute a true matrix and force the `DelegateDraw` onto its own sprite batch with the full transform applied. But, this is all theoretical, and might not work out nearly so cleanly in practice, as we recently experienced with the opacity and alpha blending stuff.

**Comment by focustense** [comment]
Made some good progress on a global transform system, although there is still a lot to do before we can even begin to look at animations. Most critical tasks:

- Make clipping work with the new transform system. Scissor rectangle are locked to the viewport and won't obey transforms, so we likely need to detect this and switch to a `RenderTarget` for clipping when `ScissorRect` can't handle it, i.e. when there's any ambient non-translation transform.
- Implement a `TransformOrigin` so that scaling, rotation, etc. can be done around the center, or any other point. Although the transform propagation does appear to work, it's always around the top-left point. Origins in XNA/MonoGame have irritating behavior, causing the entire content to shift, so they need to have their position readjusted for local transforms, and global transforms will probably have to do the "negative translation, normal transform, positive translation" sequence when the origin is not zero.
- Document the `transform` attribute and maybe also add a page or section on how transform propagation works from a user point of view.

Another task that is going to be a real headache, and may very well be deferred to a future release but still seems sort of important, long term, is to make pointer events and focus searches work with transforms. This isn't impossible; any `GlobalTransform` can be used to transform any `Point`. But combining transforms with layout might be a challenge.

**Comment by focustense** [comment]
The more I think about it, the more it seems like CSS style transitions are the right way to go here, maybe the only way to go.

Putting the target value **in** the tween data makes the system (and the data) more repetitive and harder to understand. Also, if we want to be able to handle user-configurable scenarios like animation toggles or timescales, as touched on in #79, then it is very important to differentiate between the "state" (i.e. the opacity, color, position, etc. while hovering) and the "transition". If animations are disabled, the state change should still happen, just with no transition.

To do this, the state and transition need to be held in separate places. Tweens, or transitions, are implied by the combination of the state change and the transition definition. It's the only viable way to do this _declaratively_, as opposed to programmatically.

One twist on the idea would be to have a separate set of attributes that define the target value. For example:

```html
<image sprite={@...}
       opacity="0.9"
       +hover:opacity="1"
       +transition:opacity="..."
       +hover:transform="scale: 1.2"
       +transition:transform="..." />
```

Meaning, the default opacity is 0.9 and default transform is nothing, but hovering will automatically make it opaque and scaled up, and then the transitions kick in to change the properties gradually.

One problem remaining to solve with this is that of "resting value", to return to when leaving the mutated state or when animating in a loop. It's easy in theory to save the initial state before starting a transition, and then transitioning back to the state (i.e. on pointer-leave). However, what happens if a transition is interrupted? Suddenly the starting value is different for the transition (which is good, we don't want discontinuities in the animation itself) but also for the resting state (bad, we've now lost the real original state).

Keep in mind this also has to deal with the fact _the true resting state can change_, i.e. the `opacity` itself might be a bound rather than literal value.

What I think might actually be needed here is a way to re-acquire the resting state as a function of the node graph; that is, "what would the result be if we recreated the node and/or reapplied data binding?" This gives us the true value, equivalent to the snapshot of the HTML before modifications by CSS, or even CSS traversing back from the pseudo-selector, etc.

Behaviors don't have access to that, though; they only have access to the view itself, and the data supplied to them. Therefore this might require a change in design before the two features work together properly.

**Comment by focustense** [comment]
This latest confluence of issues is proving more complicated than I first assumed, so it's time for a little rubber-ducking.

What we have is another multi-variable or multi-actor problem; concurrent (not multithreaded, but "within same update") modifications to the same state that are possibly order-dependent, previous-state-dependent and at high risk of collectively producing an inconsistent or incoherent result.

Specifically, the conflicts are between:

1. The state itself (let's say `Opacity`) being data-bound, and changing as a result of the binding;
2. One _or more_ state modifiers, for example a `hover` and `pressed` state; and
3. Tween/transition that interpolates between previous and new value.
4. Other behaviors or UI extensions making arbitrary changes to the runtime value.

All of these are acting on the _same_ property of the _same_ view, the one that is in the view tree. No matter what is done to try to wrap, decorate, etc., there must always be one current value.

To make this problem tractable, I think we have to handwave away (4), above, because to include it means we simply have no idea what a state-transition should return to. The original bound value? The modified value before transition? The value modified _during_ a transition? We don't know, and the behavior has no requirement to inform us. So if users want to mess around with that sort of thing, they should probably do it using careful ordering.

We're still left with the first 3 concerns which are all relevant and none of which can really be sacrificed. Let's consider a common set of states for a button:

- Default scale with no pointer interaction
- 1.2x default scale when hovered
- 1.1x default scale when pressed

And "default" scale is whatever the `Transform.Scale` is bound to, so if the button itself has 0.5x scale then hover state has 0.6x scale and so on. Let's also assume that scale has a transition effect; the duration and easing are irrelevant, except for being non-zero.

StardewUI's buttons don't actually have a "pressed" state, they "click" on mouse-down rather than mouse-up. But for the purposes of this example I'm just going to ignore that, as it's easy enough to conjure up similar scenarios (drags, modifier keys, etc.).

Working through the **happy path:**

- `+hover:scale` changes scale from default (for simplicity, say 1.0) to 1.2
- `+transition:scale` notices the change "somehow" and does interpolation "somehow"
- Nothing else tries to mess with scale during this period, and the transition completes
- Then there's a click, triggering `+press:scale` to change scale to 1.1
- Same as above, transition kicks in and is allowed to complete because no other state changes
- Then reverse all the above; button is released, transition completes, exits hover, transition completes again.

In this absolute best-case scenario, each behavior can simply save the state before it does whatever work it's supposed to do, and revert that state when finished. Of course there are some glaring flaws. Some obvious examples:

- The button is pressed during the hover transition; `+press:scale` no longer has a useful starting value.
- Hover state is existed while the button is still pressed; now the "starting value" for `+press:scale` has actually _changed_ and exiting press should return to the default 1.0, not the 1.2 that it saved.
- Press is not involved at all, but user moves the pointer out and quickly back in, causing a new hover to begin when the previous transition has not ended; again, we no longer have a stable starting value.

In any of the above scenarios, Bad Things Happen if a behavior just reads the current state and reverts back to it later. There _must_ be a solution for the "resting state" problem. However, even presuming we have the ability to acquire the resting state - that is, the expected value based on current bindings assuming that (4) above is discounted, i.e. that nothing else messes with the real value - it's not correct to just revert to the resting state all the time.

Consider the same happy-path case above. After exiting the pressed state, we don't want to return to the default state (scale 1.0), but rather the _hover_ state (scale 1.2). But we never actually _left_ the hover state! So the resting state of scale 1.0 is meaningless here.

This hints at what is now looking like another missing piece - a "state list" or "state stack". Suppose we maintain a dictionary of property values, each with a list or stack of what each active behavior is modifying:

- Before hover state, the `scale` entry is empty. (Note, this is really a `transform` entry, but we're simplifying.)
- On hover, we "push" `scale: 1.2`, which also sets the real scale
- On press, we push `scale: 1.1`, which again sets the real scale
- For now we're ignoring transitions, which I don't think will act on the state list (more on this later)

Now we have two states on the stack and this is where it gets interesting.

- In the happy path, we first remove the pressed state (scale 1.1), and now the hover state (scale: 1.2) is on top of the stack, so we revert to that. Then, after the pointer leaves, we remove the hover state and _nothing_ is on the stack anymore, signaling a reversion to the resting state - that is, whatever the bound/literal value is.
- Now suppose the pointer leaves but the button is still pressed. We remove the hover state from the middle of the stack, but _the pressed state is still on top_, so we don't change the value! Only when the button is released, do we remove that from the stack as well, and now, again, the stack is empty, signaling a return to the resting state, _not_ the hover state.
- One weakness of this is if the button is pressed, and then we leave and re-enter the hover state; now we have a hover state _on top_ of the pressed state which wasn't really intended, and no obvious way to prioritize the pressed state or insert the hover state underneath it instead of on top. However, this at least doesn't "corrupt" any state permanently; it flips the intended order, but when we reason through this, it's still clear that depressing will clear the pressed state (such that hover is on top, and the only active state) and that when both pressed and hover states are removed, we go back to the default bound state.

This is why I think the concept of "resting state" is still crucial. The list/stack seems to be the best and possibly only way to properly handle overlapping states, and in order for it to work, we need to know what state to return to when the stack is emptied, and it has to be trustworthy, i.e. not subject to change by some other arbitrary code.

The need to control this almost makes me wonder if behaviors shouldn't have access to the view itself at all, and instead _only_ be able to make changes through the state list. However, this seems like it might fall over in the case of _transitions_ specifically.

Transitions in particular seem hard to manage as a state on the stack because of the incredibly high likelihood of another state appearing on top of the transition, rendering the existing transition inert and orphaned. It might still be possible to work this out, I'm just not sure. I'm thinking as I write, but a hypothetical process could be:

- Check if a transition state is already at the top of the stack. If it is, then continue the transition toward the next state. ("Next state" and "previous state" can be part of the transition's own state, I think, since they are only used for the lerp itself, they don't determine anything about reverting.)
- If transition is not at the top, then _remove_ any transition in the middle of the stack as it's no longer valid - has been superseded by a new state.
- Now, if the most recent value (saved in the transition behavior) is not equal to whichever state _is_ at the top of the stack, or the resting state if the stack is empty, then start a new transition toward that value.
- When the transition completes (value reaches target), remove it from the stack.

There's a fair bit of handwaving and uncertainty here but maybe it is possible to do this without any direct access to the view, except via the state list. The required functions seem to be:

- Push a new state name + property value onto the stack
- Remove an existing state from anywhere in the stack
- Replace the value associated with a specific state (esp. for transitions) - could maybe do this via push/remove but feels clunky compared to a replace
- Peek at the top of the stack
- Get the resting value of a property, both for when the stack is empty (i.e. knowing what to transition to after all states are removed) but also for relative modifications (like scale being 1.2x the "default", not 1.2x period)
- ...and to automatically update the underlying view whenever the top of the stack changes for any given property.

Maybe this is worth a try. If it doesn't work out for transitions, I think the state list and resting state concepts could be applied to the remainder and then have transitions be a special case.

**Comment by focustense** [comment]
A little more rubber-ducking on the "resting values" concept.

It looks straightforward in theory to get the _current binding value_ for an attribute that is actually defined. `ViewNode` has access to `IViewBinding`, whose implementation has access to all the individual attribute bindings. The harder question is what to use as the default for anything _without_ an attribute binding - and this is likely to be the case for precisely the same properties that will probably use states and transitions.

For example, transforms and opacity. Transforms have a default scale of 1, not 0. Opacity has a default value of 1, not 0. Even properties like padding and margin - the "default" for that type would be `null`, since they are reference types, not `new Edges()`.

In other words, we can't simply conclude that if an attribute isn't bound, then its default is the "default for the type".

One possible way to deal with this is to _create a dummy view_, of the same type as the target view. Then, read from that dummy view to get the default values for unbound attributes. The dummy is an effectively read-only, singleton instance. Some downsides to this are:

- There's a perf cost to create the dummy view (maybe trivial if known at compile time and never used in layout?)
- Hard to guarantee that the view has a default constructor or that if it does, it would initialize to the same values as the view created via `IViewFactory`. We could maybe remedy this by actually using `IViewFactory` to create the dummy view, although...
- Doing the above imposes a lockdown on the `IViewFactory` API; while it's still fairly flexible _now_, if too many things depend on it then we can't easily extend it in the future by e.g. having attribute-dependent initializers. But maybe this is irrelevant when `IViewFactory` is part of the add-on API and therefore likely to have people depending on a stable signature regardless.

The tradeoffs don't seem very extreme. And yet, it feels like a hack and it bugs me; I just have the feeling it'll cause problems down the road. I also haven't been able to think of any clearly better options.

Maybe view descriptors could provide access to default values? I think the descriptor source generator could help with something like this, and then when we're falling back to runtime/reflection generated descriptors, _they_ could rely on creating an actual instance of the view, mostly hiding that complexity from other APIs. That way, the resting value is "bound value or descriptor default" which feels a lot saner than "bound value or read from some random object", even though it's sometimes the same underneath the abstraction.

Reflection-based descriptors could also, possibly, have some trick wherein they wait for the first instance of the view to actually be created, on the assumption that nothing should try to read the descriptor default before the view is created. This would in fact be true for behaviors. I'm not sure it's reliable as a long-term, universal assumption. Also, this forces a divorce from the `IViewFactory` since the descriptor has no knowledge of that mechanism, it's pure type related.

Perhaps a hybrid? Have _another_ abstraction, like `IViewDefaults`, which the descriptor can provide; generated descriptors can provide a generated `PrecompiledViewDefaults` at no cost, while reflection-based descriptors would return `null`, forcing the caller to go back to `IViewFactory` to create a dummy view and wrap that in a `ViewDefaults<T>`. This might be getting over-abstracted... but on the other hand it does provide an escape hatch if something goes horribly wrong with the hacky `ViewDefaults` implementation and it needs to be replaced with something else.

Not totally satisfied with the result of this mind-dump, but it's going _somewhere_, and maybe this will be good enough if no better idea presents itself.
